New resolver extras implementation

pfmoore opened the pull request:
Implemetation of extras support for the new resolver.

uranusjr commented on src/pip/_internal/resolution/resolvelib/provider.py:
```diff
@@ -38,15 +38,7 @@ def make_requirement(self, ireq):
-        #
-        # TODO: Longer term, make a proper interface for this on the candidate.
-        return getattr(c, "_ireq", None)
+        return c.get_install_requirement()
```
Oops, we’re going to get a merge conflict here…

uranusjr commented on src/pip/_internal/resolution/resolvelib/candidates.py:
```diff
@@ -120,7 +127,46 @@ def dist(self):
+
+    def get_install_requirement(self):
+        # type: () -> InstallRequirement
+        return self.base.get_install_requirement()
```
Would this return duplicated InstallRequirement objects for both `foo` and `foo[extra]`? My intuition is the return type should be `Optional[InstallRequirement]` instead and have `ExtrasCandidate.get_install_requirement()` return `None`

pfmoore replied:
Good point, I'll have more of a think about this.

pfmoore replied:
Yep, I think you're right. I've pushed a fix to do this.

uranusjr commented on src/pip/_internal/resolution/resolvelib/candidates.py:
```diff
@@ -120,7 +127,46 @@ def dist(self):
+        # Add a dependency on the exact base
+        spec = "{}=={}".format(self.base.name, self.base.version)
+        deps.append(self.base._make_install_req(spec))
+        return deps
```
Probably needs a comment or docstring explianing the implementation here for future maintainers 😛

pfmoore replied:
So you're not a fan of the "if it was hard to write, it should be hard to understand" philosophy then? 😉 

I'll add a comment here.

pfmoore replied:
Added a docstring (more like an essay really 😉)

pfmoore commented:
> Yay! Left some minor comments, but overrall this is surprisingly (for me) to read. I’m sure it took significant work to distil the implementation though!

It did, so I'm glad the effort was worthwhile! I might try to add some additional comments (maybe just a block "implementation notes" comment at the top of the file) to record the key ideas/insights (if I can remember them!!!)

> We probably need another test to ensure non-existent extras are correctly ignored.

Maybe. I'm intentionally keeping to very basic "did it work" tests at the moment - (a) to avoid getting bogged down writing tests rather than code, and (b) to avoid duplicating stuff that's in the main test suite, which we'll catch when we start really looking at the results of that with the new resolver.

But changing the existing test to install `base[add,missing]` might be enough, as well as remaining low-impact.

pfmoore commented:
OK, I've pushed fixes for the review comments, and added code to handle invalid extras by ignoring them for now. A fuller fix, to warn (and to only warn once - if we warn in get_dependencies, that method might get called multiple times) is quite a bit harder, I've added a TODO comment explaining how I think we'd need to do this and why it's not as easy as I'd like...

pfmoore commented on src/pip/_internal/resolution/resolvelib/candidates.py:
```diff
@@ -120,7 +127,46 @@ def dist(self):
+            self.base._make_install_req(str(r))
+            for r in self.base.dist.requires(self.extras)
         ]
+        # Add a dependency on the exact base
```
1. Maybe. I assumed that at a high level it doesn't matter (resolvelib solves "globally" in the end).
2. Likely via the `get_preference` method, but I don't really understand that yet 🙂

pfmoore commented on tests/functional/test_new_resolver.py:
```diff
@@ -81,3 +81,24 @@ def test_new_resolver_installs_dependencies(script):
+        "--find-links", script.scratch_path,
+        "base[add]"
+    )
+    assert_installed(script, base="0.1.0", dep="0.1.0")
```
Probably. I just went with what was easiest (I don't really understand how I'd write a new standalone YAML test). The key for me is that we just have *some* test that exercises any code we add - one install that goes through the new code path.

It's quite possible we'll throw away these tests once we are running the full test suite with the new resolver anyway.

pradyunsg replied:
It might actually be easier/nicer to bring YAML tests back into this format; and do away with all that complexity.

pfmoore replied:
I'm completely OK if you want to do that.
